pull/pull_requests.go

```diff
@@ -23,57 +23,125 @@ import (
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to list pull requests for repository %s/%s", owner, repoName)
-	}
+// GitHubClient is an interface that wraps the methods used from the github.Client.
+type GitHubClient interface {
+	ListPullRequestsWithCommit(ctx context.Context, owner, repo, sha string, opts *github.ListOptions) ([]*github.PullRequest, *github.Response, error)
+	List(ctx context.Context, owner, repo string, opts *github.PullRequestListOptions) ([]*github.PullRequest, *github.Response, error)
+}
 
+// GetOpenPullRequestsForSHA returns all open pull requests where the HEAD of the source branch
+// matches the given SHA.
+func GetOpenPullRequestsForSHA(ctx context.Context, client GitHubClient, owner, repo, sha string) ([]*github.PullRequest, error) {
+	logger := zerolog.Ctx(ctx)
 	var results []*github.PullRequest
-	for _, pr := range prs {
-		if pr.GetState() == "open" && pr.GetHead().GetSHA() == SHA {
-			results = append(results, pr)
+	opts := &github.ListOptions{PerPage: 100}
+
+	for {
+		prs, resp, err := client.ListPullRequestsWithCommit(ctx, owner, repo, sha, opts)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list pull requests for repository %s/%s", owner, repo)
+		}
+
+		for _, pr := range prs {
+			if pr.GetState() == "open" && pr.GetHead().GetSHA() == sha {
+				logger.Debug().Msgf("found open pull request with sha %s", pr.GetHead().GetSHA())
+				results = append(results, pr)
+			}
 		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opts.Page = resp.NextPage
 	}
+
 	return results, nil
 }
 
-func ListOpenPullRequestsForRef(ctx context.Context, client *github.Client, owner, repoName, ref string) ([]*github.PullRequest, error) {
+// ListOpenPullRequestsForSHA returns all open pull requests where the HEAD of the source branch
+// matches the given SHA by fetching all open PRs and filtering.
+func ListOpenPullRequestsForSHA(ctx context.Context, client GitHubClient, owner, repo, sha string) ([]*github.PullRequest, error) {
+	logger := zerolog.Ctx(ctx)
 	var results []*github.PullRequest
+	opts := &github.PullRequestListOptions{
+		State:       "open",
+		ListOptions: github.ListOptions{PerPage: 100},
+	}
+
+	for {
+		prs, resp, err := client.List(ctx, owner, repo, opts)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list pull requests for repository %s/%s", owner, repo)
+		}
+
+		for _, pr := range prs {
+			if pr.Head.GetSHA() == sha {
+				logger.Debug().Msgf("found open pull request with sha %s", pr.Head.GetSHA())
+				results = append(results, pr)
+			}
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opts.Page = resp.NextPage
+	}
+
+	return results, nil
+}
+
+// GetAllPossibleOpenPullRequestsForSHA attempts to find all open pull requests
+// associated with the given SHA using multiple methods in case we are dealing with a fork
+func GetAllPossibleOpenPullRequestsForSHA(ctx context.Context, client GitHubClient, owner, repo, sha string) ([]*github.PullRequest, error) {
```

Comment:
While the comments help, I'm wondering if we can make the differences between these functions more obvious by name. From the usage, it looks like `GetAllPossibleOpenPullRequestsForSHA`, `ListOpenPullRequestsForSHA`, and `ListOpenPullRequestsForRef` are the only functions that need to be public, so making the remaining ones private for now would help.

Beyond that, I think we should try to make it obvious which functions you should call by default and which are special cases:

* I think `GetAllPossibleOpenPullRequestsForSHA` and `ListOpenPullRequestsForRef` are the two "by default" methods, so they should have similar names. Maybe use the `Get` pattern for both of these?
* I think `ListOpenPullRequestsForSHA` is the special method that you only call directly when you know you are working with a fork and want to skip a request by not calling the "default" method. Maybe call this `ListAllOpenPullRequestsFilteredBySHA`? That might be enough to imply that there's something special about this compared to `GetOpenPullRequestsForSHA`. Open to other naming ideas too.

Reply:
Agree with what you are saying about the function names being a bit confusing. Followed your advice to make the two main functions start with Get and changed the name of the List function to make it clear that it is filtering all pull requests